File: client/src/x3dh.cpp
#include "x3dh.h"

#ifdef HAVE_SODIUM
#include <sodium.h>
#endif

#include <cstring>

namespace X3DH {

Bytes fromHex(const std::string &hex)
{
    Bytes out;
    if (hex.size() % 2) return out;
    out.reserve(hex.size() / 2);
    auto nib = [](char c) -> int {
        if (c >= '0' && c <= '9') return c - '0';
        if (c >= 'a' && c <= 'f') return c - 'a' + 10;
        if (c >= 'A' && c <= 'F') return c - 'A' + 10;
        return -1;
    };
    for (size_t i = 0; i < hex.size(); i += 2) {
        const int hi = nib(hex[i]), lo = nib(hex[i + 1]);
        if (hi < 0 || lo < 0) return Bytes();
        out.push_back(static_cast<unsigned char>((hi << 4) | lo));
    }
    return out;
}

std::string toHex(const Bytes &b)
{
    static const char *d = "0123456789abcdef";
    std::string s;
    s.reserve(b.size() * 2);
    for (unsigned char c : b) { s.push_back(d[c >> 4]); s.push_back(d[c & 15]); }
    return s;
}

#ifdef HAVE_SODIUM

namespace {

// X3DH domain separation. NOTE the info string differs from the ratchet's
// ("tamk-chat-e2ee-v1"): the two derivations must not share a label, or a secret
// from one context could stand in for the other.
const char kX3dhInfo[] = "tamk-chat-x3dh-v1";

// For Curve25519, X3DH prepends 32 bytes of 0xFF to the DH concatenation.
const unsigned char kF[32] = {
    0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff, 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
    0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff, 0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
};

// HKDF-SHA256 from HMAC, matching the helper in cryptobox.cpp.
//
// The Python reference passes a 32-byte ZERO salt where this passes nullptr.
// Those are equivalent: HMAC pads any key shorter than the block length with
// zeros, so both become the same 64 zero bytes. Verified against Python before
// this port was written rather than assumed.
void hkdf(const unsigned char *salt, size_t saltLen,
          const unsigned char *ikm, size_t ikmLen,
          const unsigned char *info, size_t infoLen,
          unsigned char *okm, size_t okmLen)
{
    unsigned char prk[crypto_auth_hmacsha256_BYTES];
    unsigned char zero[crypto_auth_hmacsha256_KEYBYTES] = {0};
    crypto_auth_hmacsha256_state st;

    if (salt == nullptr) { salt = zero; saltLen = sizeof zero; }
    crypto_auth_hmacsha256_init(&st, salt, saltLen);
    crypto_auth_hmacsha256_update(&st, ikm, ikmLen);
    crypto_auth_hmacsha256_final(&st, prk);

    unsigned char t[crypto_auth_hmacsha256_BYTES];
    size_t tLen = 0, done = 0;
    unsigned char counter = 1;
    while (done < okmLen) {
        crypto_auth_hmacsha256_init(&st, prk, sizeof prk);
        if (tLen) crypto_auth_hmacsha256_update(&st, t, tLen);
        crypto_auth_hmacsha256_update(&st, info, infoLen);
        crypto_auth_hmacsha256_update(&st, &counter, 1);
        crypto_auth_hmacsha256_final(&st, t);
        tLen = sizeof t;
        const size_t take = (okmLen - done < tLen) ? (okmLen - done) : tLen;
        memcpy(okm + done, t, take);
        done += take;
        ++counter;
    }
    sodium_memzero(prk, sizeof prk);
    sodium_memzero(t, sizeof t);
}

// X25519(our private, their public). Empty on a degenerate (low-order) result,
// which must never be fed into the KDF.
Bytes dh(const Bytes &priv, const Bytes &pub)
{
    if (priv.size() != size_t(kX25519Bytes) || pub.size() != size_t(kX25519Bytes))
        return Bytes();
    Bytes out(crypto_scalarmult_BYTES);
    if (crypto_scalarmult(out.data(), priv.data(), pub.data()) != 0) {
        sodium_memzero(out.data(), out.size());
        return Bytes();
    }
    return out;
}

// libsodium's 64-byte Ed25519 secret key (seed || public) for a 32-byte seed.
bool secretKeyFromSeed(const Bytes &seed, Bytes &sk64, Bytes &pk32)
{
    if (seed.size() != size_t(kSeedBytes)) return false;
    sk64.assign(crypto_sign_SECRETKEYBYTES, 0);
    pk32.assign(crypto_sign_PUBLICKEYBYTES, 0);
    return crypto_sign_seed_keypair(pk32.data(), sk64.data(), seed.data()) == 0;
}

Bytes kdf(const Bytes &dhConcat)
{
    Bytes ikm(kF, kF + sizeof kF);
    ikm.insert(ikm.end(), dhConcat.begin(), dhConcat.end());
    Bytes out(32);
    hkdf(nullptr, 0, ikm.data(), ikm.size(),
         reinterpret_cast<const unsigned char *>(kX3dhInfo), strlen(kX3dhInfo),
         out.data(), out.size());
    sodium_memzero(ikm.data(), ikm.size());
    return out;
}

}  // namespace

bool generateIdentity(Bytes &seedOut, Bytes &edPubOut)
{
    seedOut.assign(kSeedBytes, 0);
    randombytes_buf(seedOut.data(), seedOut.size());
    Bytes sk64;
    if (!secretKeyFromSeed(seedOut, sk64, edPubOut)) {
        sodium_memzero(seedOut.data(), seedOut.size());
        seedOut.clear();
        return false;
    }
    sodium_memzero(sk64.data(), sk64.size());
    return true;
}

Bytes identityPublic(const Bytes &seed)
{
    Bytes sk64, pk32;
    if (!secretKeyFromSeed(seed, sk64, pk32)) return Bytes();
    sodium_memzero(sk64.data(), sk64.size());
    return pk32;
}

Bytes identityToX25519Private(const Bytes &seed)
{
    Bytes sk64, pk32;
    if (!secretKeyFromSeed(seed, sk64, pk32)) return Bytes();
    Bytes out(crypto_scalarmult_SCALARBYTES);
    const int rc = crypto_sign_ed25519_sk_to_curve25519(out.data(), sk64.data());
    sodium_memzero(sk64.data(), sk64.size());
    if (rc != 0) { sodium_memzero(out.data(), out.size()); return Bytes(); }
    return out;
}

bool identityToX25519Public(const Bytes &edPub, Bytes &out)
{
    if (edPub.size() != size_t(kEdPubBytes)) return false;
    out.assign(crypto_scalarmult_BYTES, 0);
    // Returns -1 for a point not on the main subgroup. Reject rather than
    // proceed: such a bundle is malformed or hostile.
    if (crypto_sign_ed25519_pk_to_curve25519(out.data(), edPub.data()) != 0) {
        out.clear();
        return false;
    }
    return true;
}

Bytes sign(const Bytes &seed, const Bytes &message)
{
    Bytes sk64, pk32;
    if (!secretKeyFromSeed(seed, sk64, pk32)) return Bytes();
    Bytes sig(crypto_sign_BYTES);
    unsigned long long sigLen = 0;
    const int rc = crypto_sign_detached(sig.data(), &sigLen,
                                        message.data(), message.size(),
                                        sk64.data());
    sodium_memzero(sk64.data(), sk64.size());
    if (rc != 0 || sigLen != crypto_sign_BYTES) return Bytes();
    return sig;
}

bool verify(const Bytes &edPub, const Bytes &message, const Bytes &sig)
{
    if (edPub.size() != size_t(kEdPubBytes) || sig.size() != size_t(kSigBytes))
        return false;
    return crypto_sign_verify_detached(sig.data(), message.data(),
                                       message.size(), edPub.data()) == 0;
}

bool newSignedPrekey(const Bytes &ikSeed, Bytes &spkPriv, Bytes &spkPub,
                     Bytes &sigOut)
{
    spkPriv.assign(crypto_scalarmult_SCALARBYTES, 0);
    spkPub.assign(crypto_scalarmult_BYTES, 0);
    randombytes_buf(spkPriv.data(), spkPriv.size());
    crypto_scalarmult_base(spkPub.data(), spkPriv.data());
    sigOut = sign(ikSeed, spkPub);
    return !sigOut.empty();
}

bool newOneTimePrekey(Bytes &priv, Bytes &pub)
{
    priv.assign(crypto_scalarmult_SCALARBYTES, 0);
    pub.assign(crypto_scalarmult_BYTES, 0);
    randombytes_buf(priv.data(), priv.size());
    crypto_scalarmult_base(pub.data(), priv.data());
    return true;
}

bool initiatorWithEphemeral(const Bytes &aIkSeed, const Bundle &bundle,
                            const Bytes &ekPriv, InitiatorResult &out)
{
    // The signature check comes FIRST, before any DH. An unsigned or wrongly
    // signed bundle is the relay substituting keys, and computing with it would
    // establish a session with the attacker.
    if (!verify(bundle.ik, bundle.spk, bundle.spkSig)) return false;

    const Bytes aIkPub = identityPublic(aIkSeed);
    const Bytes aIkX   = identityToX25519Private(aIkSeed);
    Bytes bIkX;
    if (aIkPub.empty() || aIkX.empty()
        || !identityToX25519Public(bundle.ik, bIkX))
        return false;
    if (ekPriv.size() != size_t(kX25519Bytes)) return false;

    Bytes ekPub(crypto_scalarmult_BYTES);
    crypto_scalarmult_base(ekPub.data(), ekPriv.data());

    const Bytes dh1 = dh(aIkX,   bundle.spk);   // DH(IK_A, SPK_B)
    const Bytes dh2 = dh(ekPriv, bIkX);         // DH(EK_A, IK_B)
    const Bytes dh3 = dh(ekPriv, bundle.spk);   // DH(EK_A, SPK_B)
    if (dh1.empty() || dh2.empty() || dh3.empty()) return false;

    Bytes concat = dh1;
    concat.insert(concat.end(), dh2.begin(), dh2.end());
    concat.insert(concat.end(), dh3.begin(), dh3.end());
    if (!bundle.opk.empty()) {
        const Bytes dh4 = dh(ekPriv, bundle.opk);   // DH(EK_A, OPK_B)
        if (dh4.empty()) return false;
        concat.insert(concat.end(), dh4.begin(), dh4.end());
    }

    out.sk = kdf(concat);
    out.ik = aIkPub;
    out.ek = ekPub;
    out.ad = aIkPub;                                   // Encode(IK_A) ...
    out.ad.insert(out.ad.end(), bundle.ik.begin(), bundle.ik.end());  // ... || Encode(IK_B)
    out.opkId = bundle.opk.empty() ? -1 : bundle.opkId;

    sodium_memzero(concat.data(), concat.size());
    return !out.sk.empty();
}

bool initiator(const Bytes &aIkSeed, const Bundle &bundle, InitiatorResult &out)
{
    Bytes ekPriv(crypto_scalarmult_SCALARBYTES);
    randombytes_buf(ekPriv.data(), ekPriv.size());
    const bool ok = initiatorWithEphemeral(aIkSeed, bundle, ekPriv, out);
    sodium_memzero(ekPriv.data(), ekPriv.size());
    return ok;
}

bool responder(const Bytes &bIkSeed, const Bytes &spkPriv, const Bytes &opkPriv,
               const Bytes &ikA, const Bytes &ekA, Bytes &skOut)
{
    Bytes aIkX;
    if (!identityToX25519Public(ikA, aIkX)) return false;
    const Bytes bIkX = identityToX25519Private(bIkSeed);
    if (bIkX.empty()) return false;

    const Bytes dh1 = dh(spkPriv, aIkX);   // DH(SPK_B, IK_A)
    const Bytes dh2 = dh(bIkX,    ekA);    // DH(IK_B, EK_A)
    const Bytes dh3 = dh(spkPriv, ekA);    // DH(SPK_B, EK_A)
    if (dh1.empty() || dh2.empty() || dh3.empty()) return false;

    Bytes concat = dh1;
    concat.insert(concat.end(), dh2.begin(), dh2.end());
    concat.insert(concat.end(), dh3.begin(), dh3.end());
    if (!opkPriv.empty()) {
        const Bytes dh4 = dh(opkPriv, ekA);   // DH(OPK_B, EK_A)
        if (dh4.empty()) return false;
        concat.insert(concat.end(), dh4.begin(), dh4.end());
    }

    skOut = kdf(concat);
    sodium_memzero(concat.data(), concat.size());
    return !skOut.empty();
}

#else   // !HAVE_SODIUM

bool generateIdentity(Bytes &, Bytes &)                        { return false; }
Bytes identityPublic(const Bytes &)                            { return {}; }
Bytes identityToX25519Private(const Bytes &)                   { return {}; }
bool identityToX25519Public(const Bytes &, Bytes &)            { return false; }
Bytes sign(const Bytes &, const Bytes &)                       { return {}; }
bool verify(const Bytes &, const Bytes &, const Bytes &)       { return false; }
bool newSignedPrekey(const Bytes &, Bytes &, Bytes &, Bytes &) { return false; }
bool newOneTimePrekey(Bytes &, Bytes &)                        { return false; }
bool initiatorWithEphemeral(const Bytes &, const Bundle &, const Bytes &,
                            InitiatorResult &)                 { return false; }
bool initiator(const Bytes &, const Bundle &, InitiatorResult &) { return false; }
bool responder(const Bytes &, const Bytes &, const Bytes &, const Bytes &,
               const Bytes &, Bytes &)                         { return false; }

#endif  // HAVE_SODIUM

}  // namespace X3DH

File: tests/x3dh.py
"""
x3dh.py -- Extended Triple Diffie-Hellman (X3DH) initial key agreement.

This is the asynchronous handshake that establishes the shared secret two users
start their Double Ratchet from when one of them is offline: the recipient
publishes a prekey bundle (an identity key, a signed prekey, and a pool of
one-time prekeys), and the initiator combines several Diffie-Hellmans into one
shared secret. It is the reference the server's prekey pool
(server/prekey_store.py) and the client's bundle handling are built around.

SIGNING SCHEME -- CHANGED
-------------------------
This module previously signed the prekey with XEdDSA, implemented here with
hand-written Ed25519 arithmetic. That version carried the warning that its bytes
were "not claimed to be byte-compatible with any other XEdDSA implementation",
and it was not: libsodium's crypto_sign is RFC 8032 Ed25519, a different
construction with a different challenge hash and sign-bit convention. A C++ port
reaching for the obvious library call would have produced signatures this module
rejected, and the failure would have surfaced as an undecryptable first message
rather than as a signature error.

The identity key is now an Ed25519 SIGNING key, and the X25519 agreement key is
DERIVED from it with the standard birational map (libsodium's
crypto_sign_ed25519_pk_to_curve25519 / _sk_to_curve25519). One published key
therefore does both jobs, which is the arrangement Signal itself uses.

Both languages now use stock, well-tested implementations:

  * Python signs and verifies with `cryptography` (RFC 8032 Ed25519) and calls
    libsodium through PyNaCl only for the two conversion functions.
  * C++ signs with libsodium's crypto_sign_detached / crypto_sign_verify_detached
    and converts with crypto_sign_ed25519_*_to_curve25519.

These are byte-identical by construction, and tests/test_x3dh_vectors.py pins it
with fixed vectors rather than leaving it to trust. Roughly 150 lines of bespoke
curve arithmetic are gone with it.

DESIGN
------
  * The IDENTITY key is Ed25519 (32-byte public). Its X25519 counterpart, used
    for the Diffie-Hellmans, is derived deterministically and never published
    separately.
  * The SIGNED PREKEY and ONE-TIME PREKEYS are ordinary X25519 keys; only the
    identity needs to sign.
  * SK = HKDF-SHA256( F || DH1 || DH2 || DH3 [|| DH4] ) with the constants
    below, where DH1 = DH(IK_A, SPK_B), DH2 = DH(EK_A, IK_B),
    DH3 = DH(EK_A, SPK_B), and DH4 = DH(EK_A, OPK_B) when a one-time prekey is
    offered. IK_A and IK_B there mean the DERIVED X25519 keys.

Needs `cryptography` (X25519, Ed25519, HKDF) and `pynacl` (the conversion).
"""
from __future__ import annotations

from dataclasses import dataclass

import nacl.bindings as _sodium
from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives.serialization import (
    Encoding,
    NoEncryption,
    PrivateFormat,
    PublicFormat,
)


# --------------------------------------------------------------------------- #
# X3DH domain-separation constants
# --------------------------------------------------------------------------- #
# F: for Curve25519, X3DH prepends 32 bytes of 0xFF to the DH concatenation.
F = b"\xff" * 32
KDF_SALT = b"\x00" * 32                 # zero salt, one hash length (SHA-256)
KDF_INFO = b"tamk-chat-x3dh-v1"


# --------------------------------------------------------------------------- #
# X25519 helpers (the four DHs)
# --------------------------------------------------------------------------- #
def _priv_bytes(sk: X25519PrivateKey) -> bytes:
    return sk.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption())


def _pub_bytes(pk: X25519PublicKey) -> bytes:
    return pk.public_bytes(Encoding.Raw, PublicFormat.Raw)


def gen_key() -> tuple[bytes, bytes]:
    """A fresh X25519 keypair as (priv32, pub32)."""
    sk = X25519PrivateKey.generate()
    return _priv_bytes(sk), _pub_bytes(sk.public_key())


def _pub_from_priv(priv: bytes) -> bytes:
    return _pub_bytes(X25519PrivateKey.from_private_bytes(priv).public_key())


def _dh(priv: bytes, pub: bytes) -> bytes:
    """Raw X25519 Diffie-Hellman between a private and a public key."""
    return X25519PrivateKey.from_private_bytes(priv).exchange(
        X25519PublicKey.from_public_bytes(pub)
    )


def encode_key(pub: bytes) -> bytes:
    """Encode(PK) as used in the associated data. Both sides encode identically;
    the 32-byte public key is used directly."""
    return bytes(pub)


# --------------------------------------------------------------------------- #
# Identity: an Ed25519 signing key with a derived X25519 agreement key
# --------------------------------------------------------------------------- #
def gen_identity() -> tuple[bytes, bytes]:
    """A fresh Ed25519 identity as (seed32, public32).

    The 32-byte seed is what a device stores; libsodium's 64-byte secret key and
    the X25519 counterpart are both derived from it, so there is exactly one
    secret at rest.
    """
    sk = Ed25519PrivateKey.generate()
    seed = sk.private_bytes(Encoding.Raw, PrivateFormat.Raw, NoEncryption())
    pub = sk.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
    return seed, pub


def identity_public(seed: bytes) -> bytes:
    """The Ed25519 public key for an identity seed."""
    return (Ed25519PrivateKey.from_private_bytes(seed)
            .public_key().public_bytes(Encoding.Raw, PublicFormat.Raw))


def identity_to_x25519_private(seed: bytes) -> bytes:
    """The X25519 private key derived from an Ed25519 identity seed.

    libsodium expects its own 64-byte secret-key layout (seed || public), which
    crypto_sign_seed_keypair produces from the seed.
    """
    _pk, sk64 = _sodium.crypto_sign_seed_keypair(seed)
    return _sodium.crypto_sign_ed25519_sk_to_curve25519(sk64)


def identity_to_x25519_public(ed_pub: bytes) -> bytes:
    """The X25519 public key derived from an Ed25519 identity public key.

    Raises ValueError for a public key that is not a valid Ed25519 point, which
    is the correct response to a malformed or hostile bundle.
    """
    try:
        return _sodium.crypto_sign_ed25519_pk_to_curve25519(ed_pub)
    except Exception as exc:                      # libsodium returns -1
        raise ValueError(f"not a convertible Ed25519 public key: {exc}") from exc


# --------------------------------------------------------------------------- #
# Signing: stock RFC 8032 Ed25519, byte-identical to libsodium
# --------------------------------------------------------------------------- #
def sign(seed: bytes, message: bytes) -> bytes:
    """Detached 64-byte Ed25519 signature. Identical bytes to libsodium's
    crypto_sign_detached for the same seed and message."""
    return Ed25519PrivateKey.from_private_bytes(seed).sign(message)


def verify(ed_pub: bytes, message: bytes, signature: bytes) -> bool:
    """Verify a detached Ed25519 signature. Returns False rather than raising,
    so a caller can treat a bad bundle as data rather than as an exception."""
    if len(signature) != 64:
        return False
    try:
        Ed25519PublicKey.from_public_bytes(ed_pub).verify(signature, message)
        return True
    except (InvalidSignature, ValueError):
        return False


# --------------------------------------------------------------------------- #
# Prekeys
# --------------------------------------------------------------------------- #
def new_signed_prekey(ik_seed: bytes) -> tuple[bytes, bytes, bytes]:
    """A fresh signed prekey: (spk_priv, spk_pub, ed25519_sig_over_spk_pub).

    The prekey itself is X25519 (it is an agreement key); the signature over it
    is made with the Ed25519 identity.
    """
    spk_priv, spk_pub = gen_key()
    return spk_priv, spk_pub, sign(ik_seed, spk_pub)


def new_one_time_prekeys(n: int) -> dict[int, tuple[bytes, bytes]]:
    """A pool of `n` one-time prekeys as {opk_id: (priv, pub)}."""
    return {i: gen_key() for i in range(n)}


@dataclass
class PrekeySecrets:
    """The recipient's PRIVATE prekey material (never leaves the device)."""
    ik_seed: bytes
    spk_priv: bytes
    opks_priv: dict[int, bytes]          # {opk_id: priv}; entries are single-use


@dataclass
class PrekeyBundle:
    """The recipient's PUBLIC bundle, as served by the key directory. Mutable so
    a caller can model an exhausted pool (opk=None) or a tampered signature."""
    ik: bytes
    spk: bytes
    spk_sig: bytes
    opk: bytes | None = None
    opk_id: int | None = None


@dataclass
class InitiatorResult:
    """What the initiator derives and must transmit alongside its first message:
    the shared secret, its own identity and ephemeral publics, which one-time
    prekey it consumed, and the associated data both sides bind to."""
    sk: bytes
    ik: bytes
    ek: bytes
    opk_id: int | None
    ad: bytes


# --------------------------------------------------------------------------- #
# The agreement
# --------------------------------------------------------------------------- #
def _kdf(dh_concat: bytes) -> bytes:
    return HKDF(
        algorithm=hashes.SHA256(), length=32, salt=KDF_SALT, info=KDF_INFO
    ).derive(F + dh_concat)


def initiator_x3dh(a_ik_seed: bytes, bundle: PrekeyBundle) -> InitiatorResult:
    """Run X3DH as the initiator (the party that is online and sending first).

    Verifies the signed prekey against the bundle's identity key first -- a
    bundle whose SPK signature is forged, or was signed by a different key, is
    rejected with ValueError."""
    if not verify(bundle.ik, bundle.spk, bundle.spk_sig):
        raise ValueError("signed-prekey signature does not verify against IK")

    a_ik_pub = identity_public(a_ik_seed)
    a_ik_x = identity_to_x25519_private(a_ik_seed)
    b_ik_x = identity_to_x25519_public(bundle.ik)
    ek_priv, ek_pub = gen_key()

    dh1 = _dh(a_ik_x, bundle.spk)            # DH(IK_A, SPK_B)
    dh2 = _dh(ek_priv, b_ik_x)               # DH(EK_A, IK_B)
    dh3 = _dh(ek_priv, bundle.spk)           # DH(EK_A, SPK_B)
    dh_concat = dh1 + dh2 + dh3
    if bundle.opk is not None:
        dh_concat += _dh(ek_priv, bundle.opk)  # DH(EK_A, OPK_B)

    sk = _kdf(dh_concat)
    ad = encode_key(a_ik_pub) + encode_key(bundle.ik)
    return InitiatorResult(sk=sk, ik=a_ik_pub, ek=ek_pub,
                           opk_id=bundle.opk_id, ad=ad)


def responder_x3dh(secrets: PrekeySecrets, ik_a: bytes, ek_a: bytes,
                   opk_id: int | None) -> bytes:
    """Run X3DH as the responder, from the initiator's identity and ephemeral
    publics and the one-time prekey id it claims to have used.

    The one-time prekey is SINGLE-USE: it is removed from `secrets.opks_priv`
    here, so replaying the same opk_id raises ValueError."""
    a_ik_x = identity_to_x25519_public(ik_a)
    b_ik_x = identity_to_x25519_private(secrets.ik_seed)

    dh1 = _dh(secrets.spk_priv, a_ik_x)      # DH(SPK_B, IK_A)
    dh2 = _dh(b_ik_x, ek_a)                  # DH(IK_B, EK_A)
    dh3 = _dh(secrets.spk_priv, ek_a)        # DH(SPK_B, EK_A)
    dh_concat = dh1 + dh2 + dh3
    if opk_id is not None:
        opk_priv = secrets.opks_priv.pop(opk_id, None)   # consume it
        if opk_priv is None:
            raise ValueError(f"one-time prekey {opk_id} unknown or already used")
        dh_concat += _dh(opk_priv, ek_a)     # DH(OPK_B, EK_A)

    return _kdf(dh_concat)
